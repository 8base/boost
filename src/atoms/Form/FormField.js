import React from 'react';
import { Field } from 'react-final-form';

import { createStyledTag, createTheme } from '../../utils';

type FormFieldProps = {|
  children?: React.Node,
|};

const name = 'formField';

const theme = createTheme(name, {
  modifiers: {
  },
  defaults: {
  },
});

const StyledTag = createStyledTag(name, {});

function FormField({
  name,
  component,
  ...rest
  }: FormFieldProps) {
  return (
    <Field name={ name }>
      { ({ input }) => (
        <StyledTag tagName={ component } { ...input } { ...rest } />
      ) }
    </Field>
  );
}

export { FormField, theme };
