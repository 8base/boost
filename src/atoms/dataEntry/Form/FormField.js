// @flow

import React from 'react';

import { createStyledTag, createComponentTheme, getThemeSizes } from '../../../utils';
import * as formUtils from '../../../utils/forms';
import type { MetaType } from '../formTypes';

type FormFieldProps = {
  children?: React$Node,
  label?: string,
  stretch?: boolean,
  hideErrorLabel?: boolean,
  direction?: 'row' | 'column',
  meta: MetaType,
};

const name = 'formField';

const theme = createComponentTheme(name, {
  modifiers: {},
  defaults: {},
});

const FormFieldTag = createStyledTag(name, props => ({
  position: 'relative',
  display: 'inline-flex',
  width: props.stretch ? '100%' : 'auto',
}));

const FormFieldDirectionTag = createStyledTag(`${name}Direction`, props => ({
  display: 'inline-flex',
  flexDirection: props.direction === 'row' ? 'row-reverse' : 'column',
  alignItems: props.direction === 'row' ? 'center' : 'flex-start',
  justifyContent: props.direction === 'row' ? 'flex-end' : 'flex-start',
  width: props.stretch ? '100%' : 'auto',
}));

const ControlErrorWrapperTag = createStyledTag(`${name}ErrorWrapper`, {
  display: 'block',
  position: 'absolute',
  bottom: 0,
  height: 0,
  lineHeight: 1,
});

export const ControlErrorTag = createStyledTag(`${name}Error`, props => ({
  fontSize: '1rem',
  color: props.theme.COLORS.DANGER,
  lineHeight: 1,
  position: 'relative',
  top: '-2px',
}));

const ControlLabelTag = createStyledTag(`${name}Label`, props => ({
  marginLeft: props.direction === 'row' ? '1rem' : 0,
  fontSize: props.direction === 'row'
    ? getThemeSizes(props).MAIN_FONT_SIZE
    : getThemeSizes(props).SMALL_FONT_SIZE,
  color: props.theme.COLORS.SECONDARY_TEXT_COLOR,
  lineHeight: 2,
}));

const FormField = ({
  meta,
  label,
  children,
  direction,
  stretch,
  hideErrorLabel,
  ...rest
  }: FormFieldProps) => {
  const hasError = formUtils.hasError(meta);
  const error = formUtils.getError(meta);

  const hasLabel = !!label;

  return (
    <FormFieldTag { ...rest } stretch={ stretch } tagName="div">
      <FormFieldDirectionTag direction={ direction } stretch={ stretch } tagName="div">
        <If condition={ hasLabel }>
          <ControlLabelTag direction={ direction } tagName="div">
            { label }
          </ControlLabelTag>
        </If>
        { children }
      </FormFieldDirectionTag>
      <If condition={ hasError && !hideErrorLabel }>
        <ControlErrorWrapperTag tagName="div">
          <ControlErrorTag role="alert" tagName="span">{ error }</ControlErrorTag>
        </ControlErrorWrapperTag>
      </If>
    </FormFieldTag>
  );
};

FormField.defaultProps = {
  hideErrorLabel: false,
  stretch: true,
  direction: 'column',
};

export { FormField, theme };
