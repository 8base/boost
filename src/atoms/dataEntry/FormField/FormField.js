// @flow

import React from 'react';

import { createStyledTag, createTheme } from 'utils';
import { Label } from 'atoms/typography/Label';
import type { InputType, MetaType } from '../formTypes';

type FormFieldProps = {
  children?: React$Node,
  label?: string,
  stretch?: boolean,
  input: InputType,
  meta: MetaType,
};

const name = 'formField';

const theme = createTheme(name, {
  modifiers: { },
  defaults: {},
});

const FormFieldTag = createStyledTag(name, props => ({
  position: 'relative',
  display: 'inline-flex',
  flexDirection: 'column',

  width: props.stretch ? '100%' : 'auto',
}));
FormFieldTag.displayName = 'FormFieldTag';

const ControlErrorWrapperTag = createStyledTag(name, {
  display: 'block',
  position: 'absolute',
  bottom: 0,
  height: 0,
  lineHeight: 1,
});

const ControlErrorTag = createStyledTag(name, props => ({
  height: '1.2rem',
  color: props.theme.COLORS.RED,
  lineHeight: 2,
}));

const FormField = ({
  meta,
  label,
  children,
  ...rest
  }: FormFieldProps) => {
  const { error, touched } = meta;
  const hasError = !!error && !!touched;
  const hasLabel = !!label;

  return (
    <FormFieldTag { ...rest } tagName="div">
      <If condition={ hasLabel }>
        <Label kind="secondary">{ label }</Label>
      </If>
      { children }
      <ControlErrorWrapperTag tagName="div">
        <ControlErrorTag hasError={ hasError } tagName="div">{ error }</ControlErrorTag>
      </ControlErrorWrapperTag>
    </FormFieldTag>
  );
};

export { FormField, theme };
