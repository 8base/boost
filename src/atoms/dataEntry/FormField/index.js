// @flow

export { FormField } from './FormField';
