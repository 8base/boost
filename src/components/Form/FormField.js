// @flow

import React from 'react';
import { createThemeTag } from '../../theme/createThemeTag';

import * as formUtils from '../../utils/forms';

import type { MetaType } from '../../types';

type FormFieldProps = {
  children?: React$Node,
  label?: string,
  stretch?: boolean,
  hideErrorLabel?: boolean,
  direction?: 'row' | 'column',
  meta?: MetaType,
};

const name = 'formField';


const [FormFieldTag, themeField] = createThemeTag(name, {
  root: props => ({
    position: 'relative',
    display: 'inline-flex',
    width: props.stretch ? '100%' : 'auto',
  }),
});

const [ControlLabelTag, themeLabel] = createThemeTag(`${name}Label`, ({ COLORS, SIZES }: *) => ({
  root: props => ({
    fontSize: SIZES.OVERLINE_1,
    lineHeight: SIZES.OVERLINE_1_LH,
    color: COLORS.SECONDARY_TEXT_COLOR,
    marginLeft: props.direction === 'row' ? '8px' : 0,
  }),
}));

const [ControlErrorTag, themeError] = createThemeTag(`${name}Error`, ({ COLORS, SIZES }: *) => ({
  root: {
    position: 'relative',
    top: '-2px',

    fontSize: SIZES.OVERLINE_2,
    lineHeight: SIZES.OVERLINE_2_LH,
    color: COLORS.DANGER,
  },
}));


const [FormFieldDirectionTag, themeDirection] = createThemeTag(`${name}Direction`, {
  root: props => ({
    display: 'inline-flex',
    flexDirection: props.direction === 'row' ? 'row-reverse' : 'column',
    alignItems: props.direction === 'row' ? 'center' : 'flex-start',
    justifyContent: props.direction === 'row' ? 'flex-end' : 'flex-start',
    width: props.stretch ? '100%' : 'auto',
  }),
});

const [ControlErrorWrapperTag, themeErrorWrapper] = createThemeTag(`${name}ErrorWrapper`, {
  root: {
    display: 'block',
    position: 'absolute',
    bottom: 0,
    height: 0,
  },
});


const theme = {
  ...themeErrorWrapper,
  ...themeDirection,
  ...themeLabel,
  ...themeError,
  ...themeField,
};


const FormField = ({
  meta = {},
  label,
  children,
  hideErrorLabel,
  ...rest
  }: FormFieldProps) => {
  const hasError = formUtils.hasError(meta);
  const error = formUtils.getError(meta);

  const hasLabel = !!label;

  return (
    <FormFieldTag { ...rest } tagName="div">
      <FormFieldDirectionTag modifiers={ rest } tagName="div">
        <If condition={ hasLabel }>
          <ControlLabelTag modifiers={ rest } tagName="div">
            { label }
          </ControlLabelTag>
        </If>
        { children }
      </FormFieldDirectionTag>
      <If condition={ hasError && !hideErrorLabel }>
        <ControlErrorWrapperTag modifiers={ rest } tagName="div">
          <ControlErrorTag modifiers={ rest } role="alert" tagName="span">{ error }</ControlErrorTag>
        </ControlErrorWrapperTag>
      </If>
    </FormFieldTag>
  );
};

FormField.defaultProps = {
  hideErrorLabel: false,
  stretch: true,
  direction: 'column',
};

export { FormField, theme, ControlLabelTag, ControlErrorTag };

