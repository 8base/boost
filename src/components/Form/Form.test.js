// @flow
import React from 'react';
import { Form } from './';
import { ControlErrorTag } from './FormField';

describe('<Form />', () => {

  it('should shallow form', () => {
    const wrapperForm = shallow(<Form><div /></Form>);
    expect(wrapperForm).toMatchSnapshot();
  });


  it('should shallow form error', () => {
    const wrapperFormError = shallow(<Form.Error error="Some error" />);
    expect(wrapperFormError).toMatchSnapshot();
  });


  it('should shallow form section', () => {
    const wrapperFormSection = shallow(<Form.Section><div /></Form.Section>);
    expect(wrapperFormSection).toMatchSnapshot();
  });


  it('should shallow form section body', () => {
    const wrapperFormSectionBody = shallow(<Form.SectionBody><div /></Form.SectionBody>);
    expect(wrapperFormSectionBody).toMatchSnapshot();
  });


  it('should shallow form field', () => {
    const wrapperFormField = shallow(<Form.Field meta={{}} label="Field label" />);
    expect(wrapperFormField).toMatchSnapshot();
  });


  it('should render form components content', () => {
    const wrapper = mount(
      <BoostProvider>
        <Form>
          <Form.Error error="Some error" />
          <Form.Section>
            <Form.SectionTitle>Some section 1</Form.SectionTitle>
            <Form.SectionBody>
              <Form.Field meta={{}} label="Field label" />
            </Form.SectionBody>
          </Form.Section>
          <Form.Section>
            <Form.SectionTitle text="Some section 2" />
            <Form.SectionBody>
              <Form.Field direction="row" stretch={ false } meta={{}} label="Field label" />
            </Form.SectionBody>
          </Form.Section>
        </Form>
      </BoostProvider>,
    );

    expect(wrapper.find(Form.SectionTitle).at(0).text()).toBe('Some section 1');
    expect(wrapper.find(Form.SectionTitle).at(1).text()).toBe('Some section 2');
    expect(wrapper.find(Form.Error).text()).toBe('!Some error');
  });


  it('should render form error', () => {
    const wrapper = mount(<Form.Error error="Some error" />);
    expect(wrapper.find(Form.Error).text()).toBe('!Some error');

    wrapper.setProps({ error: '', children: 'Another error' });
    expect(wrapper.find(Form.Error).text()).toBe('!Another error');

    wrapper.setProps({ error: '', children: '' });
    expect(wrapper.find(Form.Error).props().children).toBe('');
  });


  it('should render field error', () => {
    const wrapper = mount(<Form.Field meta={{ error: 'Field error', touched: true }} />);

    expect(wrapper.find(ControlErrorTag).text()).toBe('Field error');

    wrapper.setProps({ hideErrorLabel: true });
    expect(wrapper.find(ControlErrorTag)).toHaveLength(0);
  });
});
