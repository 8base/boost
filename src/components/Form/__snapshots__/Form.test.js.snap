// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should shallow form 1`] = `
<Boost(form)
  tagName="form"
>
  <FlexLayout
    alignContent="start"
    alignItems="stretch"
    cursor="inherit"
    direction="column"
    flexWrap="nowrap"
    gap="md"
    grow={false}
    growChildren={false}
    justifyContent="start"
    offestX="none"
    offestY="none"
    stretch={false}
    tagName="div"
  >
    <div />
  </FlexLayout>
</Boost(form)>
`;

exports[`<Form /> should shallow form error 1`] = `
<Boost(formError)
  component="form"
  direction="column"
  tagName="div"
>
  <Row
    gap="md"
  >
    <Icon
      color="DANGER"
      name="Alert"
      size="md"
    />
    <Text
      color="GRAY1"
      ellipsis={false}
      weight="normal"
    >
      Some error
    </Text>
  </Row>
</Boost(formError)>
`;

exports[`<Form /> should shallow form field 1`] = `
<Boost(formField)
  direction="column"
  stretch={true}
  tagName="div"
>
  <Boost(formFieldDirection)
    modifiers={
      Object {
        "direction": "column",
        "stretch": true,
      }
    }
    tagName="div"
  >
    <Boost(formFieldLabel)
      modifiers={
        Object {
          "direction": "column",
          "stretch": true,
        }
      }
      tagName="div"
    >
      Field label
    </Boost(formFieldLabel)>
  </Boost(formFieldDirection)>
</Boost(formField)>
`;

exports[`<Form /> should shallow form section 1`] = `
<FlexLayout
  alignContent="start"
  alignItems="start"
  cursor="inherit"
  direction="column"
  flexWrap="nowrap"
  gap="sm"
  grow={false}
  growChildren={false}
  justifyContent="start"
  offestX="none"
  offestY="none"
  stretch={true}
  tagName="div"
>
  <div />
</FlexLayout>
`;

exports[`<Form /> should shallow form section body 1`] = `
<FlexLayout
  alignContent="start"
  alignItems="stretch"
  cursor="inherit"
  direction="column"
  flexWrap="nowrap"
  gap="md"
  grow={false}
  growChildren={false}
  justifyContent="start"
  offestX="none"
  offestY="none"
  stretch={true}
  tagName="div"
>
  <div />
</FlexLayout>
`;
